hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOMBucketLayoutUpgrade.java

```diff
@@ -0,0 +1,263 @@
+    omClient = clientProtocol.getOzoneManagerClient();
+
+    // create sample volume.
+    omClient.createVolume(
+        new OmVolumeArgs.Builder()
+            .setVolume(VOLUME_NAME)
+            .setOwnerName("user1")
+            .setAdminName("user1")
+            .build());
+  }
+
+  /**
+   * Shutdown MiniDFSCluster.
+   */
+  @After
+  public void shutdown() {
+    if (cluster != null) {
+      cluster.shutdown();
+    }
+  }
+
+  /**
+   * Tests that OM blocks all requests to create any buckets with a new bucket
+   * layout.
+   *
+   * @throws Exception
+   */
+  @Test
+  public void testCreateBucketWithNewLayoutsPreFinalize() throws Exception {
+    // Assert OM layout version is 'fromLayoutVersion' on deploy.
+    assertEquals(fromLayoutVersion,
+        ozoneManager.getVersionManager().getMetadataLayoutVersion());
+    assertNull(ozoneManager.getMetadataManager().getMetaTable()
+        .get(LAYOUT_VERSION_KEY));
+
+    for (BucketLayout layout : BUCKET_LAYOUTS) {
+      if (layout.isLegacy()) {
+        // We only want to test out the newly introduced layouts in this test.
+        continue;
+      }
+      try {
+        LOG.info("Creating bucket with layout {} during Pre-Finalize", layout);
+        createBucketWithLayout(layout);
+        fail("Expected to fail creating bucket with layout " + layout);
+      } catch (OMException e) {
+        // Expected exception.
+        assertTrue(e.getMessage().contains(
+            "Cluster does not have the Bucket Layout support feature " +
+                "finalized yet"));
+        LOG.info("Expected exception: " + e.getMessage());
+
+      }
+    }
+  }
+
+  /**
+   * Tests that OM allows requests to create legacy buckets.
+   *
+   * @throws Exception
+   */
+  @Test
+  public void testCreateLegacyBucketPreFinalize() throws Exception {
+    // Assert OM layout version is 'fromLayoutVersion' on deploy.
+    assertEquals(fromLayoutVersion,
+        ozoneManager.getVersionManager().getMetadataLayoutVersion());
+    assertNull(ozoneManager.getMetadataManager().getMetaTable()
+        .get(LAYOUT_VERSION_KEY));
+
+    LOG.info("Creating legacy bucket during Pre-Finalize");
+    String bucketName = createBucketWithLayout(BucketLayout.LEGACY);
+    assertEquals(
+        omClient.getBucketInfo(VOLUME_NAME, bucketName).getBucketName(),
+        bucketName);
+    assertEquals(
+        omClient.getBucketInfo(VOLUME_NAME, bucketName).getBucketLayout(),
+        BucketLayout.LEGACY);
+  }
+
+  private String createBucketWithLayout(BucketLayout bucketLayout)
+      throws Exception {
+    String bucketName = RandomStringUtils.randomAlphabetic(10).toLowerCase();
+    omClient.createBucket(
+        new OmBucketInfo.Builder()
+            .setVolumeName(VOLUME_NAME)
+            .setBucketName(bucketName)
+            .setBucketLayout(bucketLayout)
+            .build());
+
+    return bucketName;
+  }
+
+  /**
+   * Currently this is a No-Op finalization since there is only one layout
```

Comment:
This looks like an old copy/paste comment that needs to be updated.

Reply:
Fixed.